crates/criticalup-core/src/download_server_client.rs

```diff
@@ -113,13 +120,36 @@ impl DownloadServerClient {
     fn send(&self, builder: RequestBuilder) -> Result<Response, Error> {
         let req = builder.build().expect("failed to prepare the http request");
         let url = req.url().to_string();
-        let response = self
-            .client
-            .execute(req)
-            .map_err(|e| Error::DownloadServerError {
-                kind: DownloadServerError::Network(e),
-                url,
-            })?;
+
+        // Retry logic implemented here.
+        // This section implements a quick and dirty version of a retry a failed request for
+        // the max limit of `CLIENT_MAX_RETRIES` times with doubling the backoff each time
+        // starting with `CLIENT_RETRY_BACKOFF`.
+        //
+        // This logic will be superseded by changes to this code base when we move this to async.
+        let mut current_retries: u8 = 0;
+        let mut current_retry_backoff = CLIENT_RETRY_BACKOFF;
+
+        // This `try_clone()` will be `None` if request body is a stream.
+        let req_outer = req.try_clone().ok_or(Error::RequestCloningFailed)?;
+        let mut response_result = self.client.execute(req_outer);
+
+        while current_retries < CLIENT_MAX_RETRIES && response_result.is_err() {
+            // This try_clone()` will be `None` if request body is a stream.
+            let req = req.try_clone().ok_or(Error::RequestCloningFailed)?;
+
+            response_result = self.client.execute(req);
+
+            current_retries += 1;
+            current_retry_backoff *= 3;
```

Comment:
```suggestion
            current_retry_backoff *= 2;
```

Maybe make this curve slightly less steep?

100, 200, 400, 800, 1600

vs

100, 300, 900, 2700, 8100